File: Python/programmers/1844/1844.py
from collections import deque

def solution(maps):
    answer = 0
    n = len(maps[0]) # 행
    m = len(maps) # 열
    visited = [[False]*n for _ in range(m)]
    def bfs(x,y,dis) :
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        queue = deque([(x,y,1)])
        visited[y][x] = True
        while queue :
            column, row, d = queue.popleft()
            if row == m-1 and column == n-1 :
                return d

            for i in range(4) :
                next_row = row + dx[i]
                next_column = column + dy[i]
                if 0 <= next_row < m and 0 <= next_column < n :
                    if not visited[next_row][next_column] and maps[next_row][next_column] == 1 :
                        queue.append((next_column,next_row,d+1))
                        visited[next_row][next_column] = True
        return -1
    answer = bfs(0,0,1)
    
    return answer

File: C++/programmers/1844/1844.cpp
#include <vector>
#include <queue>


using namespace std;

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    int R = maps.size();
    int C = maps[0].size();
    int dr[4] = {1, 0, -1, 0};
    int dc[4] = {0, 1, 0, -1};
    queue<pair<int,int>> q;
    q.push({0,0});
    // maps[0][0] = 1;
    while(!q.empty()) {
        auto cur = q.front();
        q.pop();
        int r = cur.first;
        int c = cur.second;
        if(r == R-1 && c == C-1) return maps[r][c];
        for(int i = 0 ; i < 4 ; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            if(0 <= nr && nr < R && 0 <= nc && nc < C && maps[nr][nc] == 1) {
                maps[nr][nc] = maps[r][c] + 1;
                q.push({nr,nc});
            }
        }
    }
    return -1;
}